object_store/src/throttle.rs

```diff
@@ -0,0 +1,500 @@
+        sleep(self.wait_get_per_call).await;
+
+        // need to copy to avoid moving / referencing `self`
+        let wait_get_per_byte = self.wait_get_per_byte;
+
+        self.inner.get(location).await.map(|stream| {
+            stream
+                .then(move |bytes_result| async move {
+                    match bytes_result {
+                        Ok(bytes) => {
+                            let bytes_len: u32 = usize_to_u32_saturate(bytes.len());
+                            sleep(wait_get_per_byte * bytes_len).await;
+                            Ok(bytes)
+                        }
+                        Err(err) => Err(err),
+                    }
+                })
+                .boxed()
+        })
+    }
+
+    async fn delete(&self, location: &Self::Path) -> Result<(), Self::Error> {
+        sleep(self.wait_delete_per_call).await;
+
+        self.inner.delete(location).await
+    }
+
+    async fn list<'a>(
+        &'a self,
+        prefix: Option<&'a Self::Path>,
+    ) -> Result<BoxStream<'a, Result<Vec<Self::Path>, Self::Error>>, Self::Error> {
+        sleep(self.wait_list_per_call).await;
+
+        // need to copy to avoid moving / referencing `self`
+        let wait_list_per_entry = self.wait_list_per_entry;
+
+        self.inner.list(prefix).await.map(|stream| {
+            stream
+                .then(move |entries_result| async move {
+                    match entries_result {
+                        Ok(entries) => {
+                            let entries_len = usize_to_u32_saturate(entries.len());
+                            sleep(wait_list_per_entry * entries_len).await;
+                            Ok(entries)
+                        }
+                        Err(err) => Err(err),
+                    }
+                })
+                .boxed()
+        })
+    }
+
+    async fn list_with_delimiter(
+        &self,
+        prefix: &Self::Path,
+    ) -> Result<ListResult<Self::Path>, Self::Error> {
+        sleep(self.wait_list_with_delimiter_per_call).await;
+
+        match self.inner.list_with_delimiter(prefix).await {
+            Ok(list_result) => {
+                let entries_len = usize_to_u32_saturate(list_result.objects.len());
+                sleep(self.wait_list_with_delimiter_per_entry * entries_len).await;
+                Ok(list_result)
+            }
+            Err(err) => Err(err),
+        }
+    }
+}
+
+/// Saturated `usize` to `u32` cast.
+fn usize_to_u32_saturate(x: usize) -> u32 {
+    x.try_into().unwrap_or(u32::MAX)
+}
+
+#[cfg(test)]
+mod tests {
+    use std::iter::FromIterator;
+
+    use super::*;
+    use crate::{memory::InMemory, path::ObjectStorePath};
+    use bytes::Bytes;
+    use futures::TryStreamExt;
+    use tokio::time::Duration;
+    use tokio::time::Instant;
+
+    use crate::{
+        tests::{list_with_delimiter, put_get_delete_list},
+        ObjectStore,
+    };
```

Comment:
Here too I'd combine each crate's imports to one statement per crate:

```suggestion
    use crate::{
        memory::InMemory, 
        path::ObjectStorePath,
        tests::{list_with_delimiter, put_get_delete_list},
        ObjectStore,
    };
    use bytes::Bytes;
    use futures::TryStreamExt;
    use tokio::time::{Duration, Instant};
```